abstract solver type and behavior

dpo opened the pull request:
The objective of this PR is to introduce a simple AbstractSolver type from which other solvers can derive.

The end goal is to be able to perform efficient re-solves without reallocating a stats structure. If a user wants special storage for certain components of the stats structure (e.g., sparse vectors for bound constraints multipliers), they can allocate their own GenericExecutionStats and call `solve!(solver, model, stats)` directly.

cf: #3
related: #25
cc @sshin23

abelsiqueira commented:
What about creating `AbstractOptSolver` as well, to differentiate what `Krylov` methods can derive from. Or should we leave that for traits?

tmigot commented on src/solver.jl:
```diff
@@ -0,0 +1,9 @@
+  stats = GenericExecutionStats(:unknown, model)
+  solve!(solver, model, stats; kwargs...)
+end
+
```
```suggestion
"""
    solve!(solver, model; kwargs...)

Compute a solution of `model`.

# Arguments
- `solver::AbstractSolver`: solver structure.
- `model::AbstractNLPModel`: the model solved, see `NLPModels.jl`.
The `kwargs` are passed to the solver.

For advanced usage, it is also possible to reuse an existing `GenericExecutionStats`.
    solve!(::AbstractSolver, ::AbstractNLPModel, ::GenericExecutionStats; kwargs...)
"""
```

tmigot reviewed (commented):
That's a good idea. I made a suggestion of a docstring, feel free to adapt. You could also make a short unit test for this using `dummy_solver` to increase coverage.

tmigot commented:
If I understand correctly, the `stats` is not stored within the `solver` structure, which is what is done in Krylov.jl for instance.

geoffroyleconte commented:
Should we export `AbstractSolver` and `solve!` ?

dpo commented:
> If I understand correctly, the `stats` is not stored within the `solver` structure, which is what is done in Krylov.jl for instance.

Well, good point. We should talk about that. Krylov.jl is a bit different because there are different stats structures, so a user would be supposed to know which kind they should pass to which solver. I guess storing stats inside the solver makes that part easier for the user.

Another difference is that Krylov solvers return the solver object at the end.

From a conceptual point of view, I find it quite logical that
1. a solver object holds storage necessary to perform a *solve*;
2. a stats object holds storage necessary to represent the *result of a solve*;
3. solving returns a stats object.

But perhaps others see things differently?!

abelsiqueira commented:
> Well, good point. We should talk about that. Krylov.jl is a bit different because there are different stats structures, so a user would be supposed to know which kind they should pass to which solver. I guess storing stats inside the solver makes that part easier for the user.
> 
> Another difference is that Krylov solvers return the solver object at the end.
> 
> From a conceptual point of view, I find it quite logical that
> 
> 1. a solver object holds storage necessary to perform a _solve_;
> 2. a stats object holds storage necessary to represent the _result of a solve_;
> 3. solving returns a stats object.
> 
> But perhaps others see things differently?!

I agree with your points. But that means we change Krylov (in the future)?
About the stats being different structures, is it the case that we should allow that for Opt solvers as well? Should we have AbstractSolverOutput and Krylov outputs derived from that?

dpo commented:
> I agree with your points. But that means we change Krylov (in the future)?
About the stats being different structures, is it the case that we should allow that for Opt solvers as well? Should we have AbstractSolverOutput and Krylov outputs derived from that?

Maybe, and maybe :-). I guess Krylov and optimization solvers should be consistent. A difference is that Krylov stats don't have a `solver_specific` part. We should evaluate whether `solver_specific` makes `solve!()` type unstable or not.

dpo commented:
Rebased and added the `AbstractOptimizationSolver` type so Krylov.jl can do things differently, at least in the near future.

Adding @amontoison to the conversation.

tmigot commented on src/solver.jl:
```diff
@@ -0,0 +1,35 @@
+export AbstractSolver, solve!
```
```suggestion
export AbstractSolver, AbstractOptimizationSolver, solve!
```

tmigot commented on src/solver.jl:
```diff
@@ -0,0 +1,35 @@
+"Abstract type from which JSO solvers derive."
+abstract type AbstractSolver end
+
+abstract type AbstractOptimizationSolver <: AbstractSolver end
```
```suggestion
"Abstract type from which JSO solvers for optimization derive."
abstract type AbstractOptimizationSolver <: AbstractSolver end
```

tmigot reviewed (commented):
Just two very minor comments. The rest looks good to me. We can add unit tests with `dummy_solver` in a separate PR.

dpo commented:
Still working on the dummy solver. Trying to make it a good example to follow.

dpo commented:
Something broke: https://github.com/JuliaSmoothOptimizers/SolverCore.jl/runs/7812176580?check_suite_focus=true#step:6:222. Where is that coming from?

geoffroyleconte commented:
> Something broke: https://github.com/JuliaSmoothOptimizers/SolverCore.jl/runs/7812176580?check_suite_focus=true#step:6:222. Where is that coming from?

Could it be that some matrix was previously dense and is now sparse? Maybe `jac` in `dummy_solver`? I've had similar warnings with the SolverBenchmark tests, but the tests still passed so I though it was normal.

dpo commented:
Yes I guess ADNLPModel's `jac` now returns a sparse matrix and `A' \ b` doesn't work for sparse A: https://github.com/JuliaSparse/SparseArrays.jl/issues/115

@tmigot Could you confirm that that's due to a change in ADNLPModels?

dpo commented:
See #34.

tmigot commented:
> Yes I guess ADNLPModel's `jac` now returns a sparse matrix and `A' \ b` doesn't work for sparse A: [JuliaSparse/SparseArrays.jl#115](https://github.com/JuliaSparse/SparseArrays.jl/issues/115)
> 
> @tmigot Could you confirm that that's due to a change in ADNLPModels?

Yes, I can confirm that `jac` is now sparse in ADNLPModels (same goes for `jac_lin` and `jac_nln`).

dpo commented:
Finally got around to revising the dummy solver that's in here. If it looks ok, we could extract it and use it in the documentation.

tmigot commented on test/test_stats.jl:
```diff
@@ -66,9 +66,11 @@ function test_stats()
 
       with_logger(NullLogger()) do
-        stats = dummy_solver(nlp)
+        solve!(solver, nlp, stats)
```
```suggestion
        solve!(solver, nlp)
```

tmigot replied:
Would increase coverage sligthly

tmigot commented on test/test_stats.jl:
```diff
@@ -79,9 +81,11 @@ function test_stats()
 
       with_logger(NullLogger()) do
-        stats = dummy_solver(nlp)
+        solve!(solver, nlp, stats)
```
```suggestion
        solve!(solver, nlp)
```

tmigot reviewed (approved):
Thank you! This looks great.

tmigot commented on src/solver.jl:
```diff
@@ -0,0 +1,39 @@
+- `stats::GenericExecutionStats`: stats structure holding solution information.
+"""
+function solve!(solver::AbstractOptimizationSolver, model::AbstractNLPModel; kwargs...)
+  stats = GenericExecutionStats(:unknown, model)
```
I actually tested my suggestion to improve the coverage and got this error:
```
  MethodError: no method matching GenericExecutionStats(::Symbol, ::ADNLPModel{Float16, Vector{Float16}, Vector{Int64}})
```

tmigot commented on src/solver.jl:
```diff
@@ -0,0 +1,39 @@
+- `stats::GenericExecutionStats`: stats structure holding solution information.
+"""
+function solve!(solver::AbstractOptimizationSolver, model::AbstractNLPModel; kwargs...)
+  stats = GenericExecutionStats(:unknown, model)
```
```suggestion
  stats = GenericExecutionStats(model)
```

dpo replied:
Good catch!

tmigot commented on test/dummy_solver.jl:
```diff
@@ -1,56 +1,157 @@
+  solve!(solver, nlp, stats, args...; kwargs...)
+end
+
+function solve!(
```
```suggestion
function SolverCore.solve!(
```

dpo replied:
Hmm... that's a real drag. Thank you.

dpo commented:
This is still not ideal. If the user wants to check the iteration counter, it must be done via the `stats`. We have to decide whether the contents of `stats` can be trusted in the callback (e.g., what about the multipliers? Or the solver-specific data?)

amontoison commented:
> Rebased and added the `AbstractOptimizationSolver` type so Krylov.jl can do things differently, at least in the near future.
> 
> Adding @amontoison to the conversation.

Hi guys! Why is it relevant to have `KrylovSolver <:  AbstractOptimizationSolver`?

dpo commented:
> Hi guys! Why is it relevant to have `KrylovSolver <: AbstractOptimizationSolver`?

It's not. It should be `KrylovSolver <: Solver`, so they can behave differently than optimization solvers.

tmigot commented:
> This is still not ideal. If the user wants to check the iteration counter, it must be done via the `stats`. We have to decide whether the contents of `stats` can be trusted in the callback (e.g., what about the multipliers? Or the solver-specific data?)

Is this connected with this precise PR?

I think it should be sufficient to be precise in the docstring what fields of the stats are updated by algorithm? And add a `reset!(stats)` at the beginning of each algorithm.

dpo commented:
I agree. I should at least modify the dummy solver to illustrate how to do it.

dpo commented:
@tmigot I added a test to demo the callback feature and modified the dummy solver accordingly.

tmigot commented:
This looks great, thanks!

dpo commented:
CirrusCI is apparently having some unrelated issues. I'm going to go ahead and merge.
